### PR TITLE
Rule S4004: Collection properties should be readonly

### LINKED ISSUE #280: Rule S4004: Collection properties should be readonly
[RSPEC-4004](https://jira.sonarsource.com/browse/RSPEC-4004)

A writable collection property can be replaced by a completely different collection. Making it readonly prevents that while still allowing individual members to be set. If you want to allow the replacement of the whole collection the recommended pattern is to implement a method to remove all the elements (e.g. `System.Collections.ArrayList.Clear`) and a method to populate the collection (e.g. `System.Collections.ArrayList.AddRange`).

### PR DESCRIPTION
**Pull request by valhristov**
Fix #280

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionPropertiesShouldBeReadOnly.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class CollectionPropertiesShouldBeReadOnly : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4004";
+        private const string MessageFormat = "Make the '{0}' property 'readonly'.";
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionPropertiesShouldBeReadOnly.cs]: Please keep the message in RSPEC in sync.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/CollectionPropertiesShouldBeReadOnly.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Tests.Diagnostics
+{
+    public class Program
+    {
+        private List<string> list;
+
+        public ArrayList NongenericList { get; set; } // Noncompliant {{Make the 'NongenericList' property 'readonly'.}}
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/CollectionPropertiesShouldBeReadOnly.cs]: Please add the issue location test.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionPropertiesShouldBeReadOnly.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                var propertyDeclaration = (PropertyDeclarationSyntax)c.Node;
+                var propertySymbol = c.SemanticModel.GetDeclaredSymbol(propertyDeclaration);
+
+                if (propertyDeclaration.AccessorList != null &&
+                    propertySymbol != null &&
+                    propertySymbol.Type.DerivesOrImplementsAny(collectionTypes) &&
+                    !propertySymbol.Type.DerivesOrImplementsAny(ignoredCollectionTypes) &&
+                    propertySymbol.SetMethod != null &&
+                    !privateOrInternalAccessibility.Contains(propertySymbol.GetEffectiveAccessibility()) &&
+                    !privateOrInternalAccessibility.Contains(propertySymbol.SetMethod.DeclaredAccessibility))
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionPropertiesShouldBeReadOnly.cs]: This is quite a long if condition, but very easy to follow. I like it!

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionPropertiesShouldBeReadOnly.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class CollectionPropertiesShouldBeReadOnly : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4004";
+        private const string MessageFormat = "Make the '{0}' property read-only by removing the property setter.";
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionPropertiesShouldBeReadOnly.cs]: Perhaps consider adding "or making it private"?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionPropertiesShouldBeReadOnly.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                    IsCollection(propertySymbol.Type))
+                {
+                    c.ReportDiagnostic(Diagnostic.Create(rule, propertyDeclaration.Identifier.GetLocation(),
+                        propertySymbol.Name));
+                }
+            },
+            SyntaxKind.PropertyDeclaration);
+        }
+
+        private static bool IsCollection(ITypeSymbol typeSymbol)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionPropertiesShouldBeReadOnly.cs]: I would rename to something like `IsTrackedCollectionType`

**Comment by valhristov** [comment]
@Evangelink we have "collections should not be copied in properties" and we thought about creating "collection properties should not be modifiable from outside" (e.g. expose read-only collections only)

**Pull request merged**